src/libcore/intrinsics.rs

```diff
@@ -565,8 +565,17 @@ extern "rust-intrinsic" {
     pub fn atomic_umax_rel<T>(dst: *mut T, src: T) -> T;
     pub fn atomic_umax_acqrel<T>(dst: *mut T, src: T) -> T;
     pub fn atomic_umax_relaxed<T>(dst: *mut T, src: T) -> T;
+
+    #[cfg(not(stage0))]
+    pub fn prefetch<T>(data: *const T, rw: i32, locality: i32, cache: i32);
 }
 
+    /// Empty bootstrap implementation for stage0 compilation
+    #[cfg(stage0)]
+    pub unsafe fn prefetch<T>(_data: *const T, _rw: i32, _locality: i32, _cache: i32) {
```

Comment:
I thought of adding the binary_search optimization after this patch landed. I'm sure there are other parts of the compuiler which can bebefit from prefetching.